exec: support ON expression with INNER merge and hash joins

yuzefovich opened the pull request:
**exec: support ON expresion with INNER merge join**

Supporting ON expression is simple in case of an inner merge join
since we simply need to filter out the rows that don't satisfy the
predicate. It is done as a selection operator after the merge joiner.

**exec: add support for INNER hash joiner with ON expression**

Similarly to merge joiner case, it is very simple to add support
for ON expression in case of inner hash joiner by just planning
a filter on top. Additionally, this commit introduces a test of
hash joiner operator against the processor.

Addresses: #38018.

Release note: None

yuzefovich committed:
exec: support ON expresion with INNER merge join

Supporting ON expression is simple in case of an inner merge join
since we simply need to filter out the rows that don't satisfy the
predicate. It is done as a selection operator after the merge joiner.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/39125)
<!-- Reviewable:end -->

yuzefovich commented:
I'm not sure how to test it properly. Do we have an example of how we can construct `TypedExpr` manually?

jordanlewis commented:
That was quick work!

Check out `distsqlrun/joiner_test.go` - looks like we create them with strings and have something parse them in that test style.

yuzefovich commented:
Thanks for the pointer, I added a test.

yuzefovich committed:
exec: add support for INNER hash joiner with ON expression

Similarly to merge joiner case, it is very simple to add support
for ON expression in case of inner hash joiner by just planning
a filter on top. Additionally, this commit introduces a test of
hash joiner operator against the processor.

Release note: None

yuzefovich commented:
I realized that the same pattern can be applied to hash joiner as well, so I added the support and the corresponding test to this PR.

jordanlewis reviewed (commented):
This is a good idea, but I think the optimizer might already do this sort of transformation. Playing around with the SQL shell it kind of looks like it does, but I am not sure. @justinj do you know if the optimizer always removes extra `on` filters from joins, since they're equivalent to ordinary filters that can be pushed down and so on?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39125#-:-LkyWecK7VnrvifT1ho2:b-vmxd89)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @solongordon)



<!-- Sent from Reviewable.io -->

justinj commented:
we do the inverse transformation—we merge filters on top of inner joins with the inner join itself, see [`MergeSelectInnerJoin`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/norm/rules/select.opt#L112-L137). Assuming I understand what this PR is doing it sounds correct to me!

jordanlewis reviewed (approved):
Oops, thanks for the correction Justin.

<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39125#-:-LkzTk4245DAUd2qvmUv:b-b0oxfa)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @solongordon, and @yuzefovich)

___
*[pkg/sql/distsqlrun/columnar_operators_test.go, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39125#-LkzU63RAo37tf3NMgcP:-LkzU63RAo37tf3NMgcQ:b6yo53i) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe1f7bcf4050c9890cbf0664e4e57dd12e2c97c3/pkg/sql/distsqlrun/columnar_operators_test.go#L280)):*
> ```Go
> 		comparison = "="
> 	} else {
> 		comparison = "<>"
> ```

Why different probabilities for each of these?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39125#-:-LkzdYf43vNn4P83AtI4:b-oduykh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @solongordon)

___
*[pkg/sql/distsqlrun/columnar_operators_test.go, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39125#-LkzU63RAo37tf3NMgcP:-LkzWtkr-nH7JpTt9Rji:b-g6ysz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe1f7bcf4050c9890cbf0664e4e57dd12e2c97c3/pkg/sql/distsqlrun/columnar_operators_test.go#L280)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Why different probabilities for each of these?
</blockquote></details>

I was thinking that `<` or `>` are more interesting cases, but on a second thought, probably not, so I unified probabilities.


<!-- Sent from Reviewable.io -->
